### PR TITLE
Updated Best Practices

### PR DESCRIPTION
**Pull request by clemai**
@Michadelic Could you please review my change ? :)

### CODE REVIEW

File: webapp/test/integration/TodoList.js [hunk 1]
```diff
@@ -0,0 +1,156 @@
...
+								length: iItemCount
+							}), function(oControl) {
+								var oInput = oControl.getItems()[0].getContent()[0];
+								return new PropertyStrictEquals({
+									name: "value",
+									value: sFirstItemText
+								}).isMatching(oInput);
+							}],
+							success: function () {
+								Opa5.assert.ok(true, "The table has "+iItemCount+" item, with '"+sFirstItemText+"' as first item");
```
Comment (Michadelic) [inline webapp/test/integration/TodoList.js]: minor: spaces before operands

File: webapp/test/integration/TodoList.js [hunk 2]
```diff
@@ -0,0 +1,156 @@
...
+							},
+							errorMessage: "List does not have expected entry '"+sFirstItemText+"'."
+						});
+					},
+					iShouldSeeTheFirstItemBeingCompleted: function (bSelected) {
+						bSelected = bSelected || true;
+						return this.waitFor({
+							id: sItemListId,
+							viewName: sViewName,
+							matchers:  [function(oControl) {
```
Comment (Michadelic) [inline webapp/test/integration/TodoList.js]: no array brackets needed when there is only 1 matcher

File: webapp/test/integration/TodoList.js [hunk 3]
```diff
@@ -0,0 +1,156 @@
...
+						return this.waitFor({
+							id: sSelectedLabelId,
+							viewName: sViewName,
+							matchers:  [new PropertyStrictEquals({
+									name: "text",
+									value: numberCompleted+" completed"
+								})
+							],
+							success: function () {
+								Opa5.assert.ok(true, ""+numberCompleted+" items are selected");
```
Comment (Michadelic) [inline webapp/test/integration/TodoList.js]: minor, no ""+ needed

File: webapp/test/integration/TodoListItemActions.js [hunk 4]
```diff
@@ -0,0 +1,95 @@
+sap.ui.require([
+	"sap/ui/thirdparty/qunit",
+	"sap/ui/qunit/qunit-css",
```
Comment (Michadelic) [inline webapp/test/integration/TodoListItemActions.js]: why do you need to require qunit here?
Shouldnt it rather be included on the HTML page?
Please compare with other test pages in our demo apps

File: webapp/test/unit/controller/App.controller.js [hunk 5]
```diff
@@ -0,0 +1,110 @@
...
+	"sap/ui/thirdparty/sinon-qunit"
+], function(qunit, qunitCss, qunitjunit, View, Controller, AppController, JSONModel/*, sinon, sinonQunit*/) {
+	"use strict";
+
+	QUnit.module("Test model modification", {
+
+		beforeEach : function () {
+
+			this.oAppController = new AppController();
+			this.oViewStub = new View({});
```
Comment (Michadelic) [inline webapp/test/unit/controller/App.controller.js]: ManagedObject would be even simpler as a stub

File: webapp/test/unit/controller/App.controller.js [hunk 6]
```diff
@@ -0,0 +1,110 @@
...
+			sinon.stub(Controller.prototype, "getView").returns(this.oViewStub);
+
+			this.oModelData = {
+				todos: []
+			};
+			this.oODataModelStub = new JSONModel(this.oModelData);
```
Comment (Michadelic) [inline webapp/test/unit/controller/App.controller.js]: i would inline "this.oModelData"

File: webapp/test/unit/controller/App.controller.js [hunk 7]
```diff
@@ -0,0 +1,110 @@
...
+				"completed": false
+			}],
+			completedCount: 0
+		};
+		this.oODataModelStub.setData(oModelData);
+
+
+		// stub the event
+		var oEvt = {
+			getSource: function(){}
```
Comment (Michadelic) [inline webapp/test/unit/controller/App.controller.js]: jQuery.noop is sexier :-)

File: webapp/test/unitTests.qunit.html [hunk 8]
```diff
@@ -0,0 +1,28 @@
+<!DOCTYPE html>
+<html>
+	<head>
+		<title>Unit tests for App controller</title>
+		<meta http-equiv='X-UA-Compatible' content='IE=edge'>
+		<meta charset="utf-8">
+		<script id="sap-ui-bootstrap"
+				src="../resources/sap-ui-core.js"
+				data-sap-ui-bindingSyntax="complex"
+				data-sap-ui-compatVersion="edge"
```
Comment (Michadelic) [inline webapp/test/unitTests.qunit.html]: async bootstrap missing

File: webapp/test/unitTests.qunit.html [hunk 9]
```diff
@@ -0,0 +1,28 @@
...
+				data-sap-ui-resourceroots='{"sap.ui.todo": "../../../"}'>
```
Comment (Michadelic) [inline webapp/test/unitTests.qunit.html]: are u sure about those 3x ../?

**Comment by Michadelic** [comment]
Done!

Viele Grüße,

Michael Graf

File: webapp/index.html [hunk 10]
```diff
@@ -12,14 +12,23 @@
 		data-sap-ui-libs="sap.m"
 		data-sap-ui-bindingSyntax="complex"
 		data-sap-ui-compatVersion="edge"
-		data-sap-ui-preload="async"
```
Comment (clemai) [inline webapp/index.html]: Ohh yeah, we will readd this flag. The removal makes only sense for the internal version.

File: webapp/test/OPA5Tests.html [hunk 11]
```diff
@@ -0,0 +1,28 @@
...
+	<title>Opa tests for TodoList</title>
+	<meta http-equiv='X-UA-Compatible' content='IE=edge'>
+	<meta charset="utf-8">
+
+	<script id="sap-ui-bootstrap"  type="text/javascript"
+			src="../resources/sap-ui-core.js"
+			data-sap-ui-theme="sap_belize"
+			data-sap-ui-bindingSyntax="complex"
+			data-sap-ui-compatVersion="edge"
+			data-sap-ui-resourceroots='{"sap.ui.todo": "../"}'
```
Comment (clemai) [inline webapp/test/OPA5Tests.html]: Same reason as above.

File: webapp/test/unitTests.qunit.html (hunk 8 above)
Comment (clemai) [inline webapp/test/unitTests.qunit.html]: Same reason as above.

**Comment by clemai** [comment]
@Michadelic Thanks a lot !

File: webapp/controller/App.controller.js [hunk 12]
```diff
@@ -0,0 +1,55 @@
+sap.ui.define([
+		"sap/ui/core/mvc/Controller", "sap/ui/model/json/JSONModel", "sap/ui/model/Filter", "sap/ui/model/FilterOperator"]
```
Comment (tobiasso85) [inline webapp/controller/App.controller.js]: done

File: webapp/controller/App.controller.js [hunk 13]
```diff
@@ -0,0 +1,55 @@
...
+, function(Controller, JSONModel, Filter, FilterOperator) {
+	'use strict';
+
+	return Controller.extend('todo.controller.App', {
+
+		addTodo: function () {
```
Comment (tobiasso85) [inline webapp/controller/App.controller.js]: done

File: webapp/controller/App.controller.js [hunk 14]
```diff
@@ -0,0 +1,55 @@
...
+			// add filter for search
+			var aFilters = [];
+			var sQuery = oEvt.getSource().getValue();
+			if (sQuery && sQuery.length > 0) {
+				var filter = new Filter("title", FilterOperator.Contains, sQuery);
+				aFilters.push(filter);
+			}
+
+			// update list binding
+			var oList = this.getView().byId("todoList");
```
Comment (tobiasso85) [inline webapp/controller/App.controller.js]: would leave this out for now

File: webapp/test/integration/TodoList.js [hunk 15]
```diff
@@ -0,0 +1,156 @@
...
+		'sap/ui/test/actions/Press'
+	],
+	function (Opa5, AggregationLengthEquals, AggregationContainsPropertyEqual, PropertyStrictEquals, EnterText, Press) {
+		"use strict";
+		var sViewName = "todo.view.App",
+			sAddToItemInputId = "addTodoItemInput",
+			sItemListId = "todoList",
+			sClearCompletedId = "idClearCompleted",
+			sSelectedLabelId = "idSelectedLabel"
+			;
```
Comment (tobiasso85) [inline webapp/test/integration/TodoList.js]: put semicolon in line before

File: webapp/controller/App.controller.js (hunk 14 above)
Comment (matz3) [inline webapp/controller/App.controller.js]: You don't need to implement this on your own.
There is a `byId` method on the `Controller`:
https://openui5.hana.ondemand.com/#docs/api/symbols/sap.ui.core.mvc.Controller.html#byId

Comment (clemai) [inline webapp/controller/App.controller.js]: done

File: webapp/controller/App.controller.js [hunk 16]
```diff
@@ -55,14 +57,18 @@ sap.ui.define([
 		 * Trigger search for specific items.
```
Comment (tobiasso85) [inline webapp/controller/App.controller.js]: could you update the documentation and state that, completed items cannot be removed while a filter is active (itemsRemovable flag).

File: webapp/test/integration/TodoList.js [hunk 17]
```diff
@@ -76,76 +77,78 @@ sap.ui.require([
...
-								Opa5.assert.ok(true, "The table has 1 item, with '"+sFirstItemText+"' as first item");
+							success: function() {
+								Opa5.assert.ok(true, "The table has 1 item, with '" + sLastItemText + "' as Last item");
 							},
-							errorMessage: "List does not have expected entry '"+sFirstItemText+"'."
+							errorMessage: "List does not have expected entry '" + sLastItemText + "'."
 						});
 					},
-					iShouldSeeSelectedCount: function (numberCompleted) {
+					iShouldSeeSelectedCount: function(numberCompleted) {
```
Comment (tobiasso85) [inline webapp/test/integration/TodoList.js]: could you let this integer start with "i" -> iNumberCompleted

File: webapp/controller/App.controller.js [hunk 18]
```diff
@@ -9,12 +9,12 @@ sap.ui.define([
 	return Controller.extend('todo.controller.App', {
 
 		/**
-		 * adds a new to-do item using property <code>newTodo</code>
+		 * adds a new to-do item to the list.
```
Comment (tobiasso85) [inline webapp/controller/App.controller.js]: could you update the comments to "Adds a new to-do item to the end of the list."

**Comment by clemai** [comment]
👍

**Pull request merged**